File: AI_LabAssign_04_IDS_181-115-086.cpp
#include <bits/stdc++.h>
#define ll long long int
#define INF 2e18
using namespace std;

vector<ll> graph[100000];
vector<ll> mx;
vector<ll>arr;

void dfs(ll node, ll parn, ll height[], ll vis[])
{
  if (node != parn)
  {
      arr.push_back(node);
    height[node] = 1 + height[parn];
    parn = node;
    ll c = height[node];
    mx.push_back(c);
  }
  vis[node] = 1;

  for (ll i = 0; i < graph[node].size(); i++)
  {
    ll child = graph[node][i];
    if (vis[child] == -1)
    {
      dfs(child, parn, height, vis);
    }
  }
}

int main()
{
#ifndef ONLINE_JUDGE
  freopen("input.txt", "r", stdin);
  //freopen("output.txt", "w", stdout);
#endif
  ll v, e;
  cin >> v >> e;
  ll height[v + 1];
  ll vis[v + 1];
  memset(vis, -1, sizeof(vis));
  memset(height, -1, sizeof(height));
  for (ll i = 0; i < e; i++)
  {
    ll node1, node2;
    cin >> node1 >> node2;
    graph[node1].push_back(node2);
    graph[node2].push_back(node1);
  }
  ll source;
  cin >> source;
  height[source] = 0;
  arr.push_back(source);
  mx.push_back(height[source]);
  cout << endl;
  dfs(source, source, height, vis);
  cout << endl;
  for(ll i = 0; i < 10; i++)
  {
      cout << "When depth limit: " << i << endl;
      cout << endl;
      for(ll j = 0; j < arr.size(); j++)
      {
          if(height[arr[j]] <= i)
          {
              cout << "Explored " << arr[j] << " at depth " << height[arr[j]] << endl;
          }
      }
      cout << endl;
  }
  return 0;
}

File: AI_LabAssign_04_IDS_181-115-086.py
from collections import defaultdict
import sys
from itertools import repeat

mx = []
arr = []
res = dict(zip(range(10), repeat(None))) 



def dfs(node, parent, height, vis, tree):  
  
    # calculate the level of every node  
    # if node is not equal to parent 
    if node != parent:
        arr.append(node)
        height[node] = 1 + height[parent]  
        print("Explored", node  ,"at depth ", height[node] - 1)
        parent = node
        c = height[node] - 1
        mx.append(c)
  
    # mark every node as visited  
    vis[node] = 1
  
    # iterate in the subtree  
    for it in tree[node]:  
  
        # if the node is not visited  
        if not vis[it]:  
  
            # call the dfs function  
            dfs(it, node, height, vis, tree) 
      
    
if __name__ == "__main__":
    sys.stdin = open("input.txt", "r")
    #define dictionary as a name tree
    tree = defaultdict(list)
    #to get user input vertex and edges
    v, e = map(int, input().split())
    height = [None] * (v + 1)  
    vis = [None] * (v + 1) 
    for i in range(e):
        n1, n2 = map(int, input().split())
        tree[n1].append(n2)
        tree[n2].append(n1)
    source = int(input())
    height[source] = 1
    arr.append(source)
    print("Explored", source  ,"at depth ", height[source] - 1)
    
        #call dfs function
    dfs(source, source, height, vis, tree) 


    for i in range(0, 10):
        print("When depth limit: ", i)
        print()
        for j in range(0, len(arr)):
            if height[arr[j]] - 1 <= i:
                print("Explored ", arr[j], "at depth ", height[arr[j]] - 1)
        print()




    
   


        
     
